doc/dev/transforms.rst

```diff
@@ -0,0 +1,594 @@
+Catalyst Compiler Passes
+########################
+
+..
+    TODO: add MLIR syntax highlighting to these snippets
+    TODO: add an end-to-end guide which includes compiling and using the custom pass
+
+External resources
+==================
+
+Catalyst program transformations are implemented using the MLIR framework. A good starting point
+is to understand the IR structure and how MLIR transformations are written. Refer to the resources
+below for further information beyond this document.
+
+- `The structure and elements of the MLIR program representation <https://mlir.llvm.org/docs/LangRef/>`_.
+
+- `Understand the relationship of IR objects, and how they can be used to traverse the IR <https://mlir.llvm.org/docs/Tutorials/UnderstandingTheIRStructure/>`_.
+
+- `Documentation on general MLIR transformations (or "passes") <https://mlir.llvm.org/docs/PassManagement/>`_.
+
+
+In addition, many transformations can be expressed via pattern rewriting. A pattern is a specific
+arrangement of a group of operations into a directed acyclic graph (DAG). This allows the framework
+developer to provide a generic transformation mechanism, for which the compiler developer only needs
+to provide source and replacement patterns in order to implement a transformation.
+See the following resource for information on pattern rewrites:
+
+- `Information on specific transformations called pattern rewrites <https://mlir.llvm.org/docs/PatternRewriter/>`_.
+
+
+What does Catalyst's IR look like?
+==================================
+
+Let's look at a simple example starting in Python:
+
+.. code-block:: python
+
+    def circuit(x: complex):
```

Comment:
Maybe worth to add qjit for reproduction on the user's side

Reply:
With `@qml.qnode` and the device?

Reply:
It's mostly meant as pseudo code, the IR that's shown is not faithful either.

Reply:
No worries :+1: